### PR TITLE
[EC-98] precompiled contracts

### PR DESCRIPTION
**Pull request by rtkaczyk**
If you're interested in what they are: http://ethereum.stackexchange.com/questions/440/whats-a-precompiled-contract-and-how-are-they-different-from-native-opcodes

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  sealed trait PrecompiledContract {
+    protected def exec(inputData: ByteString): ByteString
+    protected def gas(inputData: ByteString): UInt256
+
+    def run[W <: WorldStateProxy[W, S], S <: Storage[S]](context: ProgramContext[W, S]): ProgramResult[W, S] = {
+      val g = gas(context.env.inputData)
+
+      val (result, error, gasRemaining): (ByteString, Option[ProgramError], UInt256) =
+        if (g < context.startGas)
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala]: Should we allow the execution of the precompiled contract if `g==context.startGas`?

File: src/main/scala/io/iohk/ethereum/vm/UInt256.scala [hunk 2]
```diff
@@ -180,6 +180,7 @@ class UInt256 private (private val n: BigInt) extends Ordered[UInt256] {
     //add zero if odd number of digits
     val extraZero = if (hex.length % 2 == 0) "" else "0"
     s"0x$extraZero$hex"
+    n.toString
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/vm/UInt256.scala]: Minor detail: Is this necessary? Isn't `s"0x$extraZero$hex"` the hex representation of UInt256?

**Comment by ntallar** [comment]
Apart from the comment on `g==context.startGas`, LGTM!

File: src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import scala.util.Try
+
+// scalastyle:off magic.number
+object PrecompiledContracts {
+
+  val EcDsaRecAddr = Address(1)
+  val Sha256Addr = Address(2)
+  val Rip160Addr = Address(3)
+  val IdAddr = Address(4)
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala]: account address is 20 bytes, do we pad this values with zeros?
in parity they are configured as `0000000000000000000000000000000000000001`
https://github.com/paritytech/parity/blob/master/ethcore/res/ethereum/classic.json#L61

File: src/main/scala/io/iohk/ethereum/vm/UInt256.scala (hunk 2 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/UInt256.scala]: when you add `n.toString` value produced by `s"0x$extraZero$hex"` is lost

File: src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  val examples = Table[String, String](("input", "result"),
+    ("", "9c1185a5c5e9fc54612808977ee8f548b2258d31"),
+    ("a", "0bdc9d2d256b3ee9daae347be6f4dc835a467ffe"),
+    ("abc", "8eb208f7e05d987a9b044a8e98c6b087f15a0bfc"),
+    ("message digest", "5d0689ef49d2fae572b881b123a85ffa21595f36"),
+    ("abcdefghijklmnopqrstuvwxyz", "f71c27109c692c1b56bbdceb5b9d2865b3708dbc"),
+    ("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq", "12a053384a9c0c88e405a06c27dcf49ada62eb2b"),
+    ("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789", "b0e20b6e3116640286ed3a87a5713079b21f5189"),
+    ("1234567890" * 8, "9b752e45573d4b39f4dbd3323cab82bf63326bfb"),
+    ("a" * 1000000, "52783243c1697bdbe16d37f97f68f08325dc1528")
```
Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala]: maybe generator like
arbitrary_bytes => (arbitrary_bytes, ripmed(arbitrary_bytes)) ?

File: src/test/scala/io/iohk/ethereum/vm/CallOpcodesSpec.scala [hunk 5]
```diff
@@ -218,14 +215,48 @@ class CallOpcodesSpec extends WordSpec with Matchers {
...
+        result shouldEqual expected
+      }
+
+      "return 1" in {
+        call.stateOut.stack.pop._1 shouldEqual UInt256.One
+      }
+
+      "update precompiled contract's balance" in {
+        call.extBalance shouldEqual call.value
+        call.ownBalance shouldEqual fxt.initialBalance - call.value
```
Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/vm/CallOpcodesSpec.scala]: if default precompiled contract balance is 1 wei what is correct value for precompiled contract after balance change?

File: src/main/scala/io/iohk/ethereum/vm/UInt256.scala (hunk 2 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/UInt256.scala]: Thanks Adam 😉

File: src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala (hunk 3 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala]: It's a number of a fixed bit size, so yeah, just like UInt256 its byte repr includes the zeroes.

File: src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala (hunk 4 above)
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala]: Then we would be testing that `ripemd160(arbitrary_bytes) == ripemd160(arbitrary_bytes)`.
Instead I added a comment linking to the source of those examples

File: src/test/scala/io/iohk/ethereum/vm/CallOpcodesSpec.scala (hunk 5 above)
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/vm/CallOpcodesSpec.scala]: Those accounts are like any other wrt to their balance. The test used 0 as initial balance though. Updated to 1 to make it more give it a more real world feel :)

File: src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala (hunk 4 above)
Comment (adamsmo) [inline src/test/scala/io/iohk/ethereum/crypto/Ripemd160Spec.scala]: 👍

File: src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala (hunk 3 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala]: 👍

File: src/main/scala/io/iohk/ethereum/vm/UInt256.scala (hunk 2 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/UInt256.scala]: No problem Radek 😉

**Pull request merged**